src/main/scala/com/fulcrumgenomics/fasta/UpdateFastaContigNames.scala

```diff
@@ -24,70 +24,157 @@
+    *
+    * The reference sequence to output is stored as a method, so we can lazily retrieve it and not incur large memory
+    * overhead.
+    *
+    * @param info the sequence metadata to output
+    * @param sequence the reference sequence to output
+    */
+  private class OutputSequence(val info: SequenceMetadata, sequence: => ReferenceSequence) {
+    /** Writes the output sequence */
+    def write(writer: BufferedWriter, progress: ProgressLogger): Unit = {
+      val ref = sequence
+      writer.append('>').append(info.name).append('\n')
+      val bases = ref.getBases
+      var baseCounter = 0
+      forloop(from = 0, until = bases.length) { baseIdx =>
+        writer.write(bases(baseIdx))
+        progress.record(info.name, baseIdx + 1)
+        baseCounter += 1
+        if (baseCounter >= lineLength) {
+          writer.newLine()
+          baseCounter = 0
+        }
+      }
+      if (baseCounter > 0) writer.newLine()
+    }
+  }
+
   override def execute(): Unit = {
     val progress = ProgressLogger(logger, noun="bases", verb="written", unit=10e7.toInt)
     val dict     = SequenceDictionary(this.dict)
-    val refFile  = ReferenceSequenceFileFactory.getReferenceSequenceFile(input, true, true)
     val out      = Io.toWriter(output)
 
-    val srcContigs = refFile.getSequenceDictionary match {
-      case null =>
-        require(refFile.isIndexed,
-          "Reference sequence file must have a sequence dictionary or be indexed.  Try 'picard CreateSequenceDictionary' or 'samtools faidx <ref.fasta>'.")
-        new FastaSequenceIndex(ReferenceSequenceFileFactory.getFastaIndexFileName(this.input)) map(_.getContig)
-      case dict => dict.getSequences.map(_.getSequenceName)
+    val (srcDict, srcRefFile) = getDictAndFile(fasta=this.input)
+    val defaultDictAndRefFile = this.defaultContigs.map(fasta => getDictAndFile(fasta=fasta))
+
+    // Make sure the default FASTA and the input dict are the same
+    defaultDictAndRefFile.foreach { case (defaultDict, _) =>
+      require(dict.sameAs(defaultDict), "Input sequence dictionary mismatch and default FASTA mismatch")
+    }
+
+    // Get the contigs from the input FASTA to output.
+    // Note: we do not pre-load the sequences, as to preserve memory
+    val srcSequences: Seq[OutputSequence] = srcDict.iterator.flatMap { srcInfo =>
+      // Check if the sequence dictionary can rename this contig, and if not, either log or fail depending on --skip-missing
+      val dictInfo = dict.get(srcInfo.name) match {
+        case None if skipMissing => logger.warning(s"Did not find contig ${srcInfo.name} in the sequence dictionary."); None
+        case None                => throw new IllegalStateException(s"Did not find contig ${srcInfo.name} in the sequence dictionary.")
+        case Some(info)          => Some(info)
+      }
+      dictInfo.map(info => new OutputSequence(info=info, sequence=srcRefFile.getSequence(srcInfo.name)))
+    }.toSeq
```

Comment:
My habit it to use `.toIndexedSeq` to avoid `.toSeq` returning a lazy `Stream` which can make debugging more difficult.